engine/buildkit/client.go

```diff
@@ -452,6 +492,121 @@ func withDescHandlerCacheOpts(ctx context.Context, ref bkcache.ImmutableRef) con
 	})
 }
 
+func (c *Client) ListenHostToContainer(
+	ctx context.Context,
+	hostListenAddr, proto, upstream string,
+) (*session.ListenResponse, func() error, error) {
+	clientMetadata, err := engine.ClientMetadataFromContext(ctx)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get requester session ID: %s", err)
+	}
+
+	clientCaller, err := c.SessionManager.Get(ctx, clientMetadata.ClientID, false)
```

Comment:
Is it correct that it's possible to:
1. Create a `HostService` that's bound to a `Container`
2. Do a bunch of lazy ops on top of that `Container`
3. (With Zenith) Send the container as an arg to a different client part of the same server (i.e. an environment `Function`)
4. Do something that causes the whole DAG to be evaluated by that client in the env container, including the `Exec` buried in the DAG that was bound to the host service?

If so, I think using `clientMetadata.ClientID` could result in unexpected behavior, because that refers to the client making the call that results in the evaluate, not the "original" host.

There's no need to fix this right now obviously since `Functions` aren't merged, but I think just for now we should probably switch this to [`c.MainClientCaller`](https://github.com/dagger/dagger/pull/5557/files#diff-2dd6ae1c18dbfc75417790bc661ae0d224aa1040757e9e6f8752e348973018f9R53-R57), which will result in us always forwarding to the "original" client on the real host who started the whole server.

That will put us in a slightly safer position to start once we get to merging `Function`s. I took the same approach for unix sockets in the current codebase (IIRC).

It _might_ be nice to someday support binding to the "host" of an environment (i.e. I directly listen on a socket in my env code and expose that as a service), but that's very hypothetical, so defaulting to only ever using the original host feels better for now.
* nb: We have some of the foundational plumbing in place for that in `session.go` in this package, which is why the equivalent behavior works for filesync already, but there's *a lot* of other issues unique to services that will require more work.